iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperations.java

```diff
@@ -168,6 +177,26 @@ public void commit(TableMetadata base, TableMetadata metadata) {
   @SuppressWarnings("checkstyle:MissingSwitchDefault")
   @Override
   protected void doCommit(TableMetadata base, TableMetadata metadata) {
+
+    /**
+     * During table creation, the table metadata object that arrives here has the field-ids
+     * reassigned from the client supplied schema.This code block creates a new table metadata
+     * object using the client supplied schema by preserving its field-ids.
+     */
+    if (base == null && metadata.properties().get(CatalogConstants.CLIENT_TABLE_SCHEMA) != null) {
```

Comment:
why is this not implemented in https://github.com/linkedin/openhouse/blob/b754e5e9dbafb1e83061efa2bc5357658b3ef505/services/tables/src/main/java/com/linkedin/openhouse/tables/repository/impl/OpenHouseInternalRepositoryImpl.java#L92 ? 

`doCommit` is a too low level implementation to distinguish between creation vs. update

Reply:
If I make the change in the high level code it will get overridden in this bock and committed right after https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java#L197-L204

The only place where this can be safely put without getting overridden is the commit() or doCommit() (which is called internally by commit). decided to do it in doCommit() as there are additional checks in super.commit() before doCommit() is called.

Reply:
To make sure the schema object is honored from that layer, I think what needs to happen is a customized `BaseMetastoreCatalogTableBuilder` extension so that when https://github.com/linkedin/openhouse/blob/b754e5e9dbafb1e83061efa2bc5357658b3ef505/services/tables/src/main/java/com/linkedin/openhouse/tables/repository/impl/OpenHouseInternalRepositoryImpl.java#L113 is called, the customized tableBuilder's method will be invoked and pass the schema object all the way to doCommit method.  

cc @HotSushi  since similar thing is done at the client side to provide a customized `OpenHouseTableBuilder`

Reply:
That would have been good but that approach is not ideal.

The [problematic code](https://github.com/apache/iceberg/blob/fe2f593cd025223e4ab5ab41a296fb106ce3b1cf/core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java#L197) is in the middle of the create() method of the `BaseMetastoreCatalogTableBuilder ` 

https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java#L189-L207
```
public Table create() {
      TableOperations ops = newTableOps(identifier);
      if (ops.current() != null) {
        throw new AlreadyExistsException("Table already exists: %s", identifier);
      }

      String baseLocation = location != null ? location : defaultWarehouseLocation(identifier);
      tableProperties.putAll(tableOverrideProperties());
      **TableMetadata metadata =
          TableMetadata.newTableMetadata(schema, spec, sortOrder, baseLocation, tableProperties);**

      try {
        ops.commit(null, metadata);
      } catch (CommitFailedException ignored) {
        throw new AlreadyExistsException("Table was created concurrently: %s", identifier);
      }

      return new BaseTable(ops, fullTableName(name(), identifier), metricsReporter());
    }```

If we extend the BaseMetastoreCatalogTableBuilder, then we need to override the create() method entirely and replace just the problematic code while copy pasting the rest and skip calling the create() in parent class. Code maintainability becomes hard here to keep up with OSS changes. also there are private methods (tableOverrideProperties()) being called that need to be duplicated

Reply:
I see, seems the issue of putting the fix in the tableBuilder layer is this method: 
(https://github.com/apache/iceberg/blob/5e139f48505b9b622e2cf389eae2bb74d6420510/core/src/main/java/org/apache/iceberg/TableMetadata.java#L80)